### PR TITLE
BUGFIX: SD-547 Content of tooltip is overflowed out its frame

### PR DESCRIPTION
**Pull request by ptquang86**
---

[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist

- [x] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [x] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [x] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [x] Migration guide (for major changes) is mentioned in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [x] Successful `extended test - examples`
- [x] Successful `extended test - storybook`
- [x] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)
- [x] Squash commits

# Related PRs
- gdc-analytical-designer: https://github.com/gooddata/gdc-analytical-designer/pull/2363
- gdc-dashboards: https://github.com/gooddata/gdc-dashboards/pull/2139

# Related Jira tasks
- SD-547: https://jira.intgdc.com/browse/SD-547

### CODE REVIEW

**Comment by ptquang86** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1092.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4352/ : SUCCESS in 27m 34s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5567/ : SUCCESS in 12s

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1724/ : FAILURE in 40m 45s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5568/ : SUCCESS in 9s

**Comment by ptquang86** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1092.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1730/ : SUCCESS in 16m 53s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5581/ : SUCCESS in 7s

**Comment by ptquang86** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2344/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1731/ : UNSTABLE in 11m 20s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5582/ : SUCCESS in 8s

**Comment by ptquang86** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2345/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1733/ : SUCCESS in 11m 08s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5585/ : SUCCESS in 11s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4360/ : SUCCESS in 11m 40s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5584/ : SUCCESS in 11s

File: src/components/visualizations/chart/chartOptionsBuilder.ts [hunk 1]
```diff
@@ -721,13 +721,38 @@ export function getSeries(
...
+        // padding: 13px 13px 9px;
+        // border: 1px solid transparentize($bubble-light-border-base, 0.3);
+        const tableMaxWidth = chartWidth - 28 - chartPadding; // 28 = 1 + 13 + 13 + 1
+        const columnMaxWidth = tableMaxWidth / 2; // 2 columns
+
+        tableStyle = ` style="max-width: ${tableMaxWidth}px;-webkit-border-horizontal-spacing: 0;"`;
+        columnTitleStyle = ` style="max-width: ${columnMaxWidth}px;"`;
+        // .gd-viz-tooltip-table-value => padding: 0 0 0 5px;
+        columnValueStyle = ` style="max-width: ${columnMaxWidth - 5}px;"`;
+    }
```
Comment (naibinh) [inline src/components/visualizations/chart/chartOptionsBuilder.ts]: I think we can create new `/gooddata-react-components/src/constants/tooltip.ts` then use it here
```
export const MAX_CHART_WIDTH = 340;

// https://github.com/highcharts/highcharts/blob/master/ts/parts/Chart.ts#
// plotLeft = 10, marginRight = 10;
export const CHART_PADDING = 20;

// border: 1px solid transparentize($bubble-light-border-base, 0.3);
const TOOLTIP_BORDER = 1;

// .gd-viz-tooltip-content
// padding: 13px 13px 9px;
const TOOLTIP_CONTENT_PADDING = 13;

export const TOOLTIP_PADDING = TOOLTIP_BORDER + TOOLTIP_CONTENT_PADDING + TOOLTIP_CONTENT_PADDING + TOOLTIP_BORDER; // 28

export const TOOLTIP_COLUMN_NUMBER = 2;

// .gd-viz-tooltip-table-value => padding: 0 0 0 5px;
export const TOOLTIP_VALUE_PADDING = 5;
```

File: src/components/visualizations/chart/chartOptionsBuilder.ts [hunk 2]
```diff
@@ -721,13 +721,38 @@ export function getSeries(
 
 export const customEscape = (str: string) => str && escape(unescape(str));
 
-const renderTooltipHTML = (textData: string[][]): string => {
-    return `<table class="tt-values gd-viz-tooltip-table">${textData
+const renderTooltipHTML = (textData: string[][], chartType?: string, chartWidth?: number): string => {
```
[the remark this answers is not in the document]
Comment (ptquang86) [reply src/components/visualizations/chart/chartOptionsBuilder.ts]: good catch, thank you :)

File: stories/core_components/DonutChart.tsx [hunk 3]
```diff
@@ -262,4 +262,24 @@ storiesOf("Core components/DonutChart", module)
...
+        screenshotWrap(
+            <div style={{ width: 300, height: 400 }}>
+                <DonutChart
+                    projectId="storybook"
+                    measures={[
+                        (MEASURE_1 as any)
+                            .alias("Very long name that could break layout")
+                            .format(
+                                "[<0][red]$#'#,## Very long value that could break layout;[<10000][blue]$#'#,## Very long value that could break layout;[>=10000][green]$#'#,## Very long value that could break layout;[=null] Very long value that could break layout",
+                            ),
```
[the remark this answers is not in the document]
Comment (ptquang86) [reply stories/core_components/DonutChart.tsx]: updated :)

File: src/components/visualizations/chart/chartOptionsBuilder.ts [hunk 4]
```diff
@@ -771,7 +796,8 @@ export function buildTooltipFactory(
             textData[0][0] = customEscape(point.name);
         }
 
-        return renderTooltipHTML(textData);
+        const chartWidth = get(point, ["series", "chart", "plotWidth"], 0);
```
[the remark this answers is not in the document]
Comment (ptquang86) [reply src/components/visualizations/chart/chartOptionsBuilder.ts]: updated :)

File: src/components/visualizations/chart/chartOptionsBuilder.ts (hunk 1 above)
Comment (ptquang86) [reply src/components/visualizations/chart/chartOptionsBuilder.ts]: updated :)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4363/ : SUCCESS in 9m 52s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5590/ : SUCCESS in 15s

**Comment by naibinh** [comment]
LGTM

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2354/ : SUCCESS in 12m 42s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4373/ : SUCCESS in 28m 24s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1109/ : SUCCESS in 31s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/1374/ : SUCCESS in 27m 57s

**Pull request merged**